pkg/serviceaccount/legacy.go

```diff
@@ -126,6 +140,30 @@ func (v *legacyValidator) Validate(ctx context.Context, tokenData string, public
 			klog.V(4).Infof("Service account UID no longer matches %s/%s: %q != %q", namespace, serviceAccountName, string(serviceAccount.UID), serviceAccountUID)
 			return nil, fmt.Errorf("ServiceAccount UID (%s) does not match claim (%s)", serviceAccount.UID, serviceAccountUID)
 		}
+
+		if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.LegacyServiceAccountTokenTracking) {
+			for _, ref := range serviceAccount.Secrets {
+				if ref.Name == secret.Name {
+					warning.AddWarning(ctx, "", "Use tokens from the TokenRequest API or manually created secret-based tokens instead of auto-generated secret-based tokens.")
```

Comment:
This warning isn't working because the context plumbed to token authenticators is detached from the request (because we share work if multiple requests try to authenticate with the same token concurrently):

https://github.com/kubernetes/kubernetes/blob/c8edeab2340032b272a86accc8d2039bb489915a/staging/src/k8s.io/apiserver/pkg/authentication/token/cache/cached_token_authenticator.go#L179-L186

To get the warning back up to the user, we'd need to record warnings added during authentication to the cached_token_authenticator cacheRecord, then replay them.

An integration test ensuring use of auto-generated legacy tokens issued client-visible warnings would have caught this.

Reply:
the existing integration test for warnings passes. looking...

Reply:
cached_token_authenticator was not used in the added integration test. 

change to a cached authenticator and plumb down the warning recorder.